nimble/controller/src/ble_ll_cs_drbg.c

```diff
@@ -0,0 +1,681 @@
+    memcpy(nonce, drbg_ctx->nonce_v, sizeof(nonce));
+    /* Set the Transaction_Counter */
+    if (cache->last_step_count != step_count) {
+        cache->transaction_counter = 0;
+        cache->last_step_count = step_count;
+    }
+    nonce[0] += cache->transaction_counter;
+    /* Set the Transaction_Identifier */
+    nonce[1] += transaction_id;
+    /* Set the CS Step_Counter */
+    put_le16(&nonce[2], step_count + get_le16(&nonce[2]));
+
+    while (len > 0) {
+        if (cache->free_bytes > 0) {
+            /* Use bytes from previous DRBG invocation */
+
+            if (len < cache->free_bytes) {
+                rand_len = len;
+            } else {
+                rand_len = cache->free_bytes;
+            }
+
+            cache->free_bytes -= rand_len;
+            /* [0] is LSO, [15] is MSO. Return cached bytes starting from MSO. */
+            memcpy(output, cache->random_bytes + cache->free_bytes, rand_len);
+
+            len -= rand_len;
+            output += rand_len;
+        } else {
+            /* Invoke CS_DRBG to get fresh 128-bit sequence */
+            rc = ble_ll_cs_drbg_e(drbg_ctx->key, nonce, cache->random_bytes);
+            if (rc) {
+                return rc;
+            }
+
+            cache->free_bytes = sizeof(cache->random_bytes);
+            /* Increment the transaction counter */
+            ++nonce[0];
+            ++cache->transaction_counter;
+        }
+    }
+
+    return 0;
+}
+
+/** Channel Sounding random number generation function hr1 */
+int
+ble_ll_cs_drbg_rand_hr1(struct ble_ll_cs_drbg_ctx *drbg_ctx, uint16_t step_count,
+                        uint8_t transaction_id, uint8_t r, uint8_t *r_out)
+{
+    int rc;
+    uint16_t t_rand;
+    uint8_t random_bits;
+
+    if (r <= 1) {
+        *r_out = 0;
+
+        return 0;
+    }
+
+    rc = ble_ll_cs_drbg_rand(drbg_ctx, step_count, transaction_id, &random_bits, 1);
+    if (rc) {
+        return rc;
+    }
+
+    t_rand = r * random_bits;
+
+    if ((t_rand & 0xFF) < (256 % r)) {
+        rc = ble_ll_cs_drbg_rand(drbg_ctx, step_count, transaction_id,
+                                 &random_bits, 1);
+        *r_out = ((256 * random_bits * r) + t_rand) / 65536;
+    } else {
+        *r_out = t_rand / 256;
+    }
+
+    return rc;
+}
+
+int
+ble_ll_cs_drbg_shuffle_cr1(struct ble_ll_cs_drbg_ctx *drbg_ctx, uint16_t step_count,
+                           uint8_t transaction_id, uint8_t *channel_array,
+                           uint8_t *shuffled_array, uint8_t len)
+{
+    int rc;
+    uint8_t i, j;
+
+    for (i = 0; i < len; ++i) {
+        rc = ble_ll_cs_drbg_rand_hr1(drbg_ctx, step_count, transaction_id, i + 1, &j);
+        if (rc) {
+            return rc;
+        }
+
+        if (i != j) {
+            shuffled_array[i] = shuffled_array[j];
+        }
+
+        shuffled_array[j] = channel_array[i];
+    }
+
+    return 0;
+}
+
+static int
+cs_autocorrelation_score(uint32_t sequence)
+{
+    int i, k, score = 0;
```

Comment:
should those be unsigned and abs() dropped?  (at least clang seems to be complaining)